modules/apps/29-fee/ibc_middleware.go

```diff
@@ -322,7 +322,32 @@ func (im IBCMiddleware) OnTimeoutPacket(
 
 // OnChanUpgradeInit implements the IBCModule interface
 func (im IBCMiddleware) OnChanUpgradeInit(ctx sdk.Context, portID, channelID string, order channeltypes.Order, connectionHops []string, sequence uint64, version, previousVersion string) (string, error) {
-	return im.app.OnChanUpgradeInit(ctx, portID, channelID, order, connectionHops, sequence, version, previousVersion)
+	var versionMetadata types.Metadata
+	if err := types.ModuleCdc.UnmarshalJSON([]byte(version), &versionMetadata); err != nil {
+		// Since it is valid for fee version to not be specified, the above middleware version may be for a middleware
+		// lower down in the stack. Thus, if it is not a fee version we pass the entire version string onto the underlying
+		// application.
+		return im.app.OnChanUpgradeInit(ctx, portID, channelID, order, connectionHops, sequence, version, previousVersion)
+	}
+
+	if versionMetadata.FeeVersion != types.Version {
+		return "", errorsmod.Wrapf(types.ErrInvalidVersion, "expected %s, got %s", types.Version, versionMetadata.FeeVersion)
+	}
+
+	appVersion, err := im.app.OnChanUpgradeInit(ctx, portID, channelID, order, connectionHops, sequence, versionMetadata.AppVersion, previousVersion)
+	if err != nil {
+		return "", err
+	}
+
+	versionMetadata.AppVersion = appVersion
+	versionBz, err := types.ModuleCdc.MarshalJSON(&versionMetadata)
+	if err != nil {
+		return "", err
+	}
+
+	im.keeper.SetFeeEnabled(ctx, portID, channelID)
```

Comment:
do we already want to setFeeEnabled at this step?

Reply:
we set is as enabled in `OnChanOpenInit` so I expected the upgrade handlers should follow suit, setting fee enabled on INIT and TRY, and then asserting that the handshake must finish as fee enabled in ACK and CONFIRM.

Reply:
ah okay, this makes sense. thanks!

Reply:
do you know the reasoning we set fee enabled on INIT and TRY already for `OnChanOpenInit`?

Reply:
I'm not 100% sure, it's been that way since @AdityaSripal added the initial implementation in #307. 
I imagine it could be useful if the underlying app is doing something in the channel handshake handlers (INIT/TRY) based on whether fee is enabled or not.